### ISSUE TITLE
Possible synchronizationQueue mis-configuration?

### ISSUE THREAD

**Issue by ptsochantaris** [Alamofire/AlamofireImage#331]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/AlamofireImage/blob/master/CONTRIBUTING.md).

## What did you do?

Called `AutoPurgingImageCache.image(withIdentifier: ...)` from multiple threads.

## What did you expect to happen?

The result to be returned without any threading issues, as the method internally uses a thread-safe synchronisation queue.

## What happened instead?

A thread race condition was detected by the Xcode thread sanitiser.

## Alamofire Environment

**Alamofire version:** 4.7.3 / **AlamofireImage version:** 3.4.1
**Xcode version:** 10.0
**Swift version:** 4.2
**Platform(s) running AlamofireImage:** iOS
**macOS version running Xcode:** Mojave

I think there may be a small but important misconfiguration of the `syncrhonizationQueue` in the file `ImageCache.swift`.

That queue, in this specific instance, is initialised as a concurrent queue - which on the surface seems to make it pointless, since it won't actually synchronise anything if it is concurrent. This is the code I'm referring to:

```
self.synchronizationQueue = {
  let name = String(format: "org.alamofire.autopurgingimagecache-%08x%08x", arc4random(), arc4random())
  return DispatchQueue(label: name, attributes: .concurrent)
}()
```

In contrast the `synchronizationQueue` in `ImageDownloader.swift` seems to be correctly initialised as a serial queue, and hence will order requests to the shared resource atomically:

```
private let synchronizationQueue: DispatchQueue = {
  let name = String(format: "org.alamofire.imagedownloader.synchronizationqueue-%08x%08x", arc4random(), arc4random())
  return DispatchQueue(label: name) // <-- no "concurrent" attribute
}()
```

Please accept my apologies and feel free to close the issue in case I have misunderstood the purpose of that dispatch queue, but I thought I'd rather mention it and be wrong, than not mention it at all :)

**Comment by ptsochantaris** [comment]
Hi @jshier and thanks for getting back to me. Please find attached a screenshot of the sanitiser as well as the output from the console, slightly redacting our internal stack calls, but I hope the info you need should all still be there.

In order to reproduce the issue I basically have to shift the calls to `image(withIdentifier: ...)` for the shared cache object inside a background thread and then it's a question of time until the threads "cross paths", which implies the synchronisation isn't quite working the way its expected. Basically if you create a shared AutoPurgingImageCache and throw calls at it from different threads, this is what will happen.

As a workaround I'm firing these off from the main thread right now, and as such they seem to arrive "in order", even inside the synchronisation queue, and not trigger the problem.

Happy to provide any more info you may need!

<img width="1297" alt="sanitiser" src="https://user-images.githubusercontent.com/521581/47780202-da738400-dcf2-11e8-9759-be60987db5ec.png">

Here is the console output:

[sanitiser-output.txt](https://github.com/Alamofire/AlamofireImage/files/2533671/sanitiser-output.txt)

**Comment by zSOLz** [comment]
If we want to keep possibility to read images from multiple threads and do it in parallel (single writer - multiple readers pattern) we should use `synchronizationQueue.sync(flags: [.barrier]) {}` basically everywhere we update inner data (all cases except `image(withIdentifier:)` method).

**Comment by jshier** [comment]
This issue was correct and has been fixed in 3.6.0 and 4.0.0-beta.6.
